File: 02-cleanup-picture/LightXCleanupPicture.swift
/**
 * LightX Cleanup Picture API Integration - Swift
 * 
 * This implementation provides a complete integration with LightX API v2
 * for image cleanup functionality using mask-based object removal.
 */

import Foundation
import UIKit

// MARK: - Data Models

struct CleanupUploadImageResponse: Codable {
    let statusCode: Int
    let message: String
    let body: CleanupUploadImageBody
}

struct CleanupUploadImageBody: Codable {
    let uploadImage: String
    let imageUrl: String
    let size: Int
}

struct CleanupPictureResponse: Codable {
    let statusCode: Int
    let message: String
    let body: CleanupPictureBody
}

struct CleanupPictureBody: Codable {
    let orderId: String
    let maxRetriesAllowed: Int
    let avgResponseTimeInSec: Int
    let status: String
}

struct CleanupOrderStatusResponse: Codable {
    let statusCode: Int
    let message: String
    let body: CleanupOrderStatusBody
}

struct CleanupOrderStatusBody: Codable {
    let orderId: String
    let status: String
    let output: String?
}

// MARK: - LightX Cleanup API Client

class LightXCleanupAPI {
    
    // MARK: - Properties
    
    private let apiKey: String
    private let baseURL = "https://api.lightxeditor.com/external/api"
    private let maxRetries = 5
    private let retryInterval: TimeInterval = 3.0 // seconds
    
    // MARK: - Initialization
    
    init(apiKey: String) {
        self.apiKey = apiKey
    }
    
    // MARK: - Public Methods
    
    /**
     * Upload image to LightX servers
     * - Parameters:
     *   - imageData: Image data to upload
     *   - contentType: MIME type (image/jpeg or image/png)
     * - Returns: Final image URL
     */
    func uploadImage(imageData: Data, contentType: String = "image/jpeg") async throws -> String {
        let fileSize = imageData.count
        
        guard fileSize <= 5242880 else { // 5MB limit
            throw LightXCleanupError.imageSizeExceeded
        }
        
        // Step 1: Get upload URL
        let uploadURL = try await getUploadURL(fileSize: fileSize, contentType: contentType)
        
        // Step 2: Upload image to S3
        try await uploadToS3(uploadURL: uploadURL, imageData: imageData, contentType: contentType)
        
        print("✅ Image uploaded successfully")
        return uploadURL.imageUrl
    }
    
    /**
     * Upload multiple images (original and mask)
     * - Parameters:
     *   - originalImageData: Original image data
     *   - maskImageData: Mask image data
     *   - contentType: MIME type
     * - Returns: Tuple of (originalURL, maskURL)
     */
    func uploadImages(originalImageData: Data, maskImageData: Data, 
                     contentType: String = "image/jpeg") async throws -> (String, String) {
        print("📤 Uploading original image...")
        let originalURL = try await uploadImage(imageData: originalImageData, contentType: contentType)
        
        print("📤 Uploading mask image...")
        let maskURL = try await uploadImage(imageData: maskImageData, contentType: contentType)
        
        return (originalURL, maskURL)
    }
    
    /**
     * Cleanup picture using mask
     * - Parameters:
     *   - imageURL: URL of the original image
     *   - maskedImageURL: URL of the mask image
     * - Returns: Order ID for tracking
     */
    func cleanupPicture(imageURL: String, maskedImageURL: String) async throws -> String {
        let endpoint = "\(baseURL)/v1/cleanup-picture"
        
        guard let url = URL(string: endpoint) else {
            throw LightXCleanupError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        
        let payload = [
            "imageUrl": imageURL,
            "maskedImageUrl": maskedImageURL
        ]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw LightXCleanupError.networkError
        }
        
        let cleanupResponse = try JSONDecoder().decode(CleanupPictureResponse.self, from: data)
        
        guard cleanupResponse.statusCode == 2000 else {
            throw LightXCleanupError.apiError(cleanupResponse.message)
        }
        
        let orderInfo = cleanupResponse.body
        
        print("📋 Order created: \(orderInfo.orderId)")
        print("🔄 Max retries allowed: \(orderInfo.maxRetriesAllowed)")
        print("⏱️  Average response time: \(orderInfo.avgResponseTimeInSec) seconds")
        print("📊 Status: \(orderInfo.status)")
        
        return orderInfo.orderId
    }
    
    /**
     * Check order status
     * - Parameter orderID: Order ID to check
     * - Returns: Order status and results
     */
    func checkOrderStatus(orderID: String) async throws -> CleanupOrderStatusBody {
        let endpoint = "\(baseURL)/v1/order-status"
        
        guard let url = URL(string: endpoint) else {
            throw LightXCleanupError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        
        let payload = ["orderId": orderID]
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw LightXCleanupError.networkError
        }
        
        let statusResponse = try JSONDecoder().decode(CleanupOrderStatusResponse.self, from: data)
        
        guard statusResponse.statusCode == 2000 else {
            throw LightXCleanupError.apiError(statusResponse.message)
        }
        
        return statusResponse.body
    }
    
    /**
     * Wait for order completion with automatic retries
     * - Parameter orderID: Order ID to monitor
     * - Returns: Final result with output URL
     */
    func waitForCompletion(orderID: String) async throws -> CleanupOrderStatusBody {
        var attempts = 0
        
        while attempts < maxRetries {
            do {
                let status = try await checkOrderStatus(orderID: orderID)
                
                print("🔄 Attempt \(attempts + 1): Status - \(status.status)")
                
                switch status.status {
                case "active":
                    print("✅ Picture cleanup completed successfully!")
                    if let output = status.output {
                        print("🖼️  Output image: \(output)")
                    }
                    return status
                    
                case "failed":
                    throw LightXCleanupError.processingFailed
                    
                case "init":
                    attempts += 1
                    if attempts < maxRetries {
                        print("⏳ Waiting \(retryInterval) seconds before next check...")
                        try await Task.sleep(nanoseconds: UInt64(retryInterval * 1_000_000_000))
                    }
                    
                default:
                    attempts += 1
                    if attempts < maxRetries {
                        try await Task.sleep(nanoseconds: UInt64(retryInterval * 1_000_000_000))
                    }
                }
                
            } catch {
                attempts += 1
                if attempts >= maxRetries {
                    throw error
                }
                print("⚠️  Error on attempt \(attempts), retrying...")
                try await Task.sleep(nanoseconds: UInt64(retryInterval * 1_000_000_000))
            }
        }
        
        throw LightXCleanupError.maxRetriesReached
    }
    
    /**
     * Complete workflow: Upload images and cleanup picture
     * - Parameters:
     *   - originalImageData: Original image data
     *   - maskImageData: Mask image data
     *   - contentType: MIME type
     * - Returns: Final result with output URL
     */
    func processCleanup(originalImageData: Data, maskImageData: Data, 
                       contentType: String = "image/jpeg") async throws -> CleanupOrderStatusBody {
        print("🚀 Starting LightX Cleanup Picture API workflow...")
        
        // Step 1: Upload both images
        print("📤 Uploading images...")
        let (originalURL, maskURL) = try await uploadImages(
            originalImageData: originalImageData, 
            maskImageData: maskImageData, 
            contentType: contentType
        )
        print("✅ Original image uploaded: \(originalURL)")
        print("✅ Mask image uploaded: \(maskURL)")
        
        // Step 2: Cleanup picture
        print("🧹 Cleaning up picture...")
        let orderID = try await cleanupPicture(imageURL: originalURL, maskedImageURL: maskURL)
        
        // Step 3: Wait for completion
        print("⏳ Waiting for processing to complete...")
        let result = try await waitForCompletion(orderID: orderID)
        
        return result
    }
    
    /**
     * Create a simple mask from coordinates (utility function)
     * - Parameters:
     *   - width: Image width
     *   - height: Image height
     *   - coordinates: Array of CGRect for white areas
     * - Returns: Mask image data
     */
    func createMaskFromCoordinates(width: Int, height: Int, coordinates: [CGRect]) -> Data? {
        print("🎭 Creating mask from coordinates...")
        print("Image dimensions: \(width)x\(height)")
        print("White areas: \(coordinates)")
        
        let size = CGSize(width: width, height: height)
        let renderer = UIGraphicsImageRenderer(size: size)
        
        let maskImage = renderer.image { context in
            // Fill with black background
            UIColor.black.setFill()
            context.fill(CGRect(origin: .zero, size: size))
            
            // Draw white rectangles for areas to remove
            UIColor.white.setFill()
            for rect in coordinates {
                context.fill(rect)
            }
        }
        
        return maskImage.jpegData(compressionQuality: 1.0)
    }
    
    // MARK: - Private Methods
    
    private func getUploadURL(fileSize: Int, contentType: String) async throws -> CleanupUploadImageBody {
        let endpoint = "\(baseURL)/v2/uploadImageUrl"
        
        guard let url = URL(string: endpoint) else {
            throw LightXCleanupError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        
        let payload = [
            "uploadType": "imageUrl",
            "size": fileSize,
            "contentType": contentType
        ] as [String: Any]
        
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw LightXCleanupError.networkError
        }
        
        let uploadResponse = try JSONDecoder().decode(CleanupUploadImageResponse.self, from: data)
        
        guard uploadResponse.statusCode == 2000 else {
            throw LightXCleanupError.apiError(uploadResponse.message)
        }
        
        return uploadResponse.body
    }
    
    private func uploadToS3(uploadURL: CleanupUploadImageBody, imageData: Data, contentType: String) async throws {
        guard let url = URL(string: uploadURL.uploadImage) else {
            throw LightXCleanupError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        request.httpBody = imageData
        
        let (_, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw LightXCleanupError.uploadFailed
        }
    }
}

// MARK: - Error Handling

enum LightXCleanupError: Error, LocalizedError {
    case invalidURL
    case networkError
    case apiError(String)
    case imageSizeExceeded
    case uploadFailed
    case processingFailed
    case maxRetriesReached
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .networkError:
            return "Network error occurred"
        case .apiError(let message):
            return "API Error: \(message)"
        case .imageSizeExceeded:
            return "Image size exceeds 5MB limit"
        case .uploadFailed:
            return "Image upload failed"
        case .processingFailed:
            return "Picture cleanup processing failed"
        case .maxRetriesReached:
            return "Maximum retry attempts reached"
        }
    }
}

// MARK: - Example Usage

@MainActor
class LightXCleanupExample {
    
    static func runExample() async {
        do {
            // Initialize with your API key
            let lightx = LightXCleanupAPI(apiKey: "YOUR_API_KEY_HERE")
            
            // Load original image data (replace with your image loading logic)
            guard let originalImage = UIImage(named: "original_image"),
                  let originalImageData = originalImage.jpegData(compressionQuality: 0.8) else {
                print("❌ Failed to load original image")
                return
            }
            
            // Option 1: Use existing mask image
            guard let maskImage = UIImage(named: "mask_image"),
                  let maskImageData = maskImage.jpegData(compressionQuality: 1.0) else {
                print("❌ Failed to load mask image")
                return
            }
            
            // Process the cleanup
            let result = try await lightx.processCleanup(
                originalImageData: originalImageData,
                maskImageData: maskImageData,
                contentType: "image/jpeg"
            )
            
            print("🎉 Final result:")
            print("Order ID: \(result.orderId)")
            print("Status: \(result.status)")
            if let output = result.output {
                print("Output: \(output)")
            }
            
            // Option 2: Create mask from coordinates
            // let maskData = lightx.createMaskFromCoordinates(
            //     width: 800, height: 600,
            //     coordinates: [
            //         CGRect(x: 100, y: 100, width: 200, height: 150),  // Area to remove
            //         CGRect(x: 400, y: 300, width: 100, height: 100)   // Another area to remove
            //     ]
            // )
            // 
            // if let maskData = maskData {
            //     let result = try await lightx.processCleanup(
            //         originalImageData: originalImageData,
            //         maskImageData: maskData,
            //         contentType: "image/jpeg"
            //     )
            //     print("🎉 Final result with generated mask:")
            //     print("Order ID: \(result.orderId)")
            //     print("Status: \(result.status)")
            //     if let output = result.output {
            //         print("Output: \(output)")
            //     }
            // }
            
        } catch {
            print("❌ Example failed: \(error.localizedDescription)")
        }
    }
}

File: 02-cleanup-picture/LightXCleanupPicture.kt
/**
 * LightX Cleanup Picture API Integration - Kotlin
 * 
 * This implementation provides a complete integration with LightX API v2
 * for image cleanup functionality using mask-based object removal.
 */

import kotlinx.coroutines.*
import kotlinx.serialization.*
import kotlinx.serialization.json.*
import java.io.*
import java.net.*
import java.net.http.*
import java.time.Duration
import java.awt.image.BufferedImage
import java.awt.Color
import java.awt.Graphics2D
import javax.imageio.ImageIO

// MARK: - Data Classes

@Serializable
data class CleanupUploadImageResponse(
    val statusCode: Int,
    val message: String,
    val body: CleanupUploadImageBody
)

@Serializable
data class CleanupUploadImageBody(
    val uploadImage: String,
    val imageUrl: String,
    val size: Int
)

@Serializable
data class CleanupPictureResponse(
    val statusCode: Int,
    val message: String,
    val body: CleanupPictureBody
)

@Serializable
data class CleanupPictureBody(
    val orderId: String,
    val maxRetriesAllowed: Int,
    val avgResponseTimeInSec: Int,
    val status: String
)

@Serializable
data class CleanupOrderStatusResponse(
    val statusCode: Int,
    val message: String,
    val body: CleanupOrderStatusBody
)

@Serializable
data class CleanupOrderStatusBody(
    val orderId: String,
    val status: String,
    val output: String? = null
)

// MARK: - Coordinate Data Class

data class MaskCoordinate(
    val x: Int,
    val y: Int,
    val width: Int,
    val height: Int
)

// MARK: - LightX Cleanup API Client

class LightXCleanupAPI(private val apiKey: String) {
    
    companion object {
        private const val BASE_URL = "https://api.lightxeditor.com/external/api"
        private const val MAX_RETRIES = 5
        private const val RETRY_INTERVAL = 3000L // milliseconds
        private const val MAX_FILE_SIZE = 5242880L // 5MB
    }
    
    private val httpClient = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(30))
        .build()
    
    private val json = Json { ignoreUnknownKeys = true }
    
    /**
     * Upload image to LightX servers
     * @param imageFile File containing the image data
     * @param contentType MIME type (image/jpeg or image/png)
     * @return Final image URL
     */
    suspend fun uploadImage(imageFile: File, contentType: String = "image/jpeg"): String {
        val fileSize = imageFile.length()
        
        if (fileSize > MAX_FILE_SIZE) {
            throw LightXCleanupException("Image size exceeds 5MB limit")
        }
        
        // Step 1: Get upload URL
        val uploadUrl = getUploadURL(fileSize, contentType)
        
        // Step 2: Upload image to S3
        uploadToS3(uploadUrl.uploadImage, imageFile, contentType)
        
        println("✅ Image uploaded successfully")
        return uploadUrl.imageUrl
    }
    
    /**
     * Upload multiple images (original and mask)
     * @param originalImageFile Original image file
     * @param maskImageFile Mask image file
     * @param contentType MIME type
     * @return Pair of (originalURL, maskURL)
     */
    suspend fun uploadImages(originalImageFile: File, maskImageFile: File, 
                           contentType: String = "image/jpeg"): Pair<String, String> {
        println("📤 Uploading original image...")
        val originalUrl = uploadImage(originalImageFile, contentType)
        
        println("📤 Uploading mask image...")
        val maskUrl = uploadImage(maskImageFile, contentType)
        
        return Pair(originalUrl, maskUrl)
    }
    
    /**
     * Cleanup picture using mask
     * @param imageUrl URL of the original image
     * @param maskedImageUrl URL of the mask image
     * @return Order ID for tracking
     */
    suspend fun cleanupPicture(imageUrl: String, maskedImageUrl: String): String {
        val endpoint = "$BASE_URL/v1/cleanup-picture"
        
        val requestBody = buildJsonObject {
            put("imageUrl", imageUrl)
            put("maskedImageUrl", maskedImageUrl)
        }
        
        val request = HttpRequest.newBuilder()
            .uri(URI.create(endpoint))
            .header("Content-Type", "application/json")
            .header("x-api-key", apiKey)
            .POST(HttpRequest.BodyPublishers.ofString(requestBody.toString()))
            .build()
        
        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        
        if (response.statusCode() != 200) {
            throw LightXCleanupException("Network error: ${response.statusCode()}")
        }
        
        val cleanupResponse = json.decodeFromString<CleanupPictureResponse>(response.body())
        
        if (cleanupResponse.statusCode != 2000) {
            throw LightXCleanupException("Cleanup picture request failed: ${cleanupResponse.message}")
        }
        
        val orderInfo = cleanupResponse.body
        
        println("📋 Order created: ${orderInfo.orderId}")
        println("🔄 Max retries allowed: ${orderInfo.maxRetriesAllowed}")
        println("⏱️  Average response time: ${orderInfo.avgResponseTimeInSec} seconds")
        println("📊 Status: ${orderInfo.status}")
        
        return orderInfo.orderId
    }
    
    /**
     * Check order status
     * @param orderId Order ID to check
     * @return Order status and results
     */
    suspend fun checkOrderStatus(orderId: String): CleanupOrderStatusBody {
        val endpoint = "$BASE_URL/v1/order-status"
        
        val requestBody = buildJsonObject {
            put("orderId", orderId)
        }
        
        val request = HttpRequest.newBuilder()
            .uri(URI.create(endpoint))
            .header("Content-Type", "application/json")
            .header("x-api-key", apiKey)
            .POST(HttpRequest.BodyPublishers.ofString(requestBody.toString()))
            .build()
        
        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        
        if (response.statusCode() != 200) {
            throw LightXCleanupException("Network error: ${response.statusCode()}")
        }
        
        val statusResponse = json.decodeFromString<CleanupOrderStatusResponse>(response.body())
        
        if (statusResponse.statusCode != 2000) {
            throw LightXCleanupException("Status check failed: ${statusResponse.message}")
        }
        
        return statusResponse.body
    }
    
    /**
     * Wait for order completion with automatic retries
     * @param orderId Order ID to monitor
     * @return Final result with output URL
     */
    suspend fun waitForCompletion(orderId: String): CleanupOrderStatusBody {
        var attempts = 0
        
        while (attempts < MAX_RETRIES) {
            try {
                val status = checkOrderStatus(orderId)
                
                println("🔄 Attempt ${attempts + 1}: Status - ${status.status}")
                
                when (status.status) {
                    "active" -> {
                        println("✅ Picture cleanup completed successfully!")
                        status.output?.let { println("🖼️  Output image: $it") }
                        return status
                    }
                    "failed" -> throw LightXCleanupException("Picture cleanup failed")
                    "init" -> {
                        attempts++
                        if (attempts < MAX_RETRIES) {
                            println("⏳ Waiting ${RETRY_INTERVAL / 1000} seconds before next check...")
                            delay(RETRY_INTERVAL)
                        }
                    }
                    else -> {
                        attempts++
                        if (attempts < MAX_RETRIES) {
                            delay(RETRY_INTERVAL)
                        }
                    }
                }
                
            } catch (e: Exception) {
                attempts++
                if (attempts >= MAX_RETRIES) {
                    throw e
                }
                println("⚠️  Error on attempt $attempts, retrying...")
                delay(RETRY_INTERVAL)
            }
        }
        
        throw LightXCleanupException("Maximum retry attempts reached")
    }
    
    /**
     * Complete workflow: Upload images and cleanup picture
     * @param originalImageFile Original image file
     * @param maskImageFile Mask image file
     * @param contentType MIME type
     * @return Final result with output URL
     */
    suspend fun processCleanup(
        originalImageFile: File, 
        maskImageFile: File, 
        contentType: String = "image/jpeg"
    ): CleanupOrderStatusBody {
        println("🚀 Starting LightX Cleanup Picture API workflow...")
        
        // Step 1: Upload both images
        println("📤 Uploading images...")
        val (originalUrl, maskUrl) = uploadImages(originalImageFile, maskImageFile, contentType)
        println("✅ Original image uploaded: $originalUrl")
        println("✅ Mask image uploaded: $maskUrl")
        
        // Step 2: Cleanup picture
        println("🧹 Cleaning up picture...")
        val orderId = cleanupPicture(originalUrl, maskUrl)
        
        // Step 3: Wait for completion
        println("⏳ Waiting for processing to complete...")
        val result = waitForCompletion(orderId)
        
        return result
    }
    
    /**
     * Create a simple mask from coordinates (utility function)
     * @param width Image width
     * @param height Image height
     * @param coordinates List of MaskCoordinate for white areas
     * @return File containing the mask image
     */
    fun createMaskFromCoordinates(width: Int, height: Int, coordinates: List<MaskCoordinate>): File {
        println("🎭 Creating mask from coordinates...")
        println("Image dimensions: ${width}x${height}")
        println("White areas: $coordinates")
        
        // Create black background
        val maskImage = BufferedImage(width, height, BufferedImage.TYPE_INT_RGB)
        val graphics = maskImage.createGraphics()
        
        // Fill with black
        graphics.color = Color.BLACK
        graphics.fillRect(0, 0, width, height)
        
        // Draw white rectangles for areas to remove
        graphics.color = Color.WHITE
        for (coord in coordinates) {
            graphics.fillRect(coord.x, coord.y, coord.width, coord.height)
        }
        
        graphics.dispose()
        
        // Save mask to file
        val maskFile = File("generated_mask.png")
        ImageIO.write(maskImage, "PNG", maskFile)
        println("✅ Mask created and saved: ${maskFile.absolutePath}")
        
        return maskFile
    }
    
    // MARK: - Private Methods
    
    private suspend fun getUploadURL(fileSize: Long, contentType: String): CleanupUploadImageBody {
        val endpoint = "$BASE_URL/v2/uploadImageUrl"
        
        val requestBody = buildJsonObject {
            put("uploadType", "imageUrl")
            put("size", fileSize)
            put("contentType", contentType)
        }
        
        val request = HttpRequest.newBuilder()
            .uri(URI.create(endpoint))
            .header("Content-Type", "application/json")
            .header("x-api-key", apiKey)
            .POST(HttpRequest.BodyPublishers.ofString(requestBody.toString()))
            .build()
        
        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        
        if (response.statusCode() != 200) {
            throw LightXCleanupException("Network error: ${response.statusCode()}")
        }
        
        val uploadResponse = json.decodeFromString<CleanupUploadImageResponse>(response.body())
        
        if (uploadResponse.statusCode != 2000) {
            throw LightXCleanupException("Upload URL request failed: ${uploadResponse.message}")
        }
        
        return uploadResponse.body
    }
    
    private suspend fun uploadToS3(uploadUrl: String, imageFile: File, contentType: String) {
        val request = HttpRequest.newBuilder()
            .uri(URI.create(uploadUrl))
            .header("Content-Type", contentType)
            .PUT(HttpRequest.BodyPublishers.ofFile(imageFile.toPath()))
            .build()
        
        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        
        if (response.statusCode() != 200) {
            throw LightXCleanupException("Image upload failed: ${response.statusCode()}")
        }
    }
    
    /**
     * Close the HTTP client
     */
    fun close() {
        httpClient.close()
    }
}

// MARK: - Exception Classes

class LightXCleanupException(message: String) : Exception(message)

// MARK: - Example Usage

object LightXCleanupExample {
    
    @JvmStatic
    suspend fun main(args: Array<String>) {
        try {
            // Initialize with your API key
            val lightx = LightXCleanupAPI("YOUR_API_KEY_HERE")
            
            // Option 1: Process with existing images
            val originalImageFile = File("path/to/original-image.jpg")
            val maskImageFile = File("path/to/mask-image.png")
            
            if (!originalImageFile.exists() || !maskImageFile.exists()) {
                println("❌ Image files not found")
                return
            }
            
            val result = lightx.processCleanup(
                originalImageFile = originalImageFile,
                maskImageFile = maskImageFile,
                contentType = "image/jpeg"
            )
            
            println("🎉 Final result:")
            println("Order ID: ${result.orderId}")
            println("Status: ${result.status}")
            result.output?.let { println("Output: $it") }
            
            // Option 2: Create mask from coordinates and process
            // val maskFile = lightx.createMaskFromCoordinates(
            //     width = 800, height = 600,
            //     coordinates = listOf(
            //         MaskCoordinate(100, 100, 200, 150),  // Area to remove
            //         MaskCoordinate(400, 300, 100, 100)   // Another area to remove
            //     )
            // )
            // 
            // val resultWithGeneratedMask = lightx.processCleanup(
            //     originalImageFile = originalImageFile,
            //     maskImageFile = maskFile,
            //     contentType = "image/jpeg"
            // )
            // 
            // println("🎉 Final result with generated mask:")
            // println("Order ID: ${resultWithGeneratedMask.orderId}")
            // println("Status: ${resultWithGeneratedMask.status}")
            // resultWithGeneratedMask.output?.let { println("Output: $it") }
            
        } catch (e: Exception) {
            println("❌ Example failed: ${e.message}")
        }
    }
}

// MARK: - Coroutine Extension Functions

/**
 * Extension function to run the example in a coroutine scope
 */
fun runLightXCleanupExample() {
    runBlocking {
        LightXCleanupExample.main(emptyArray())
    }
}
